File: Python3/1118.py
n=int(input())
father = [i for i in range(10010)]
forest = set()
isTree = {}
def findfather(x):
    a = x
    while x != father[x]:
        x = father[x]
    while a != father[a]:
        z = a
        a = father[a]
        father[z] = x
    return x

def Union(a,b):
    fa = findfather(a)
    fb = findfather(b)
    if fa != fb:
        father[fa] = fb

for _ in range(n):
    _,*birds = map(int,input().split())
    forest.add(birds[0])
    for bird in birds[1:]:
        Union(birds[0],bird)
        forest.add(bird)
cnt = 0
for bird in forest:
    fa = findfather(bird)
    if isTree.get(fa) is None:
        isTree[fa] = True
        cnt += 1
print(cnt,len(forest))
k = int(input())

for _ in range(k):
    a,b = map(int,input().split())
    fa = findfather(a)
    fb = findfather(b)
    if fa != fb:
        print("No")
    else:
        print("Yes")

File: c++/1118.cpp
//
//  main.cpp
//  test
//
//  Created by 李寻欢 on 2019/10/15.
//  Copyright © 2019 李寻欢. All rights reserved.
//
#include <stdio.h>
#include <iostream>
#include <string>
#include <map>
#include <cmath>
#include <algorithm>
#include <stack>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <string.h>
#include <set>
#include <deque>
using namespace std;
const int MAXN = 1e4+10;
int n,k;
int father[MAXN];
bool isTree[MAXN] ={false};
set<int> forest;
int findfather(int x){
    int a = x;
    while (a != father[a]) {
        a = father[a];
    }
    while (x != father[x]) {
        int z = x;
        x = father[x];
        father[z] = a;
    }
    return x;
}

void Union(int a,int b){
    int fa = findfather(a);
    int fb = findfather(b);
    if(fa != fb) father[fa] = fb;
}



int main(){
    scanf("%d",&n);
    int birds,cur;
    for(int i=0;i<MAXN;++i) father[i] = i;
    for(int i=0;i<n;++i){
        scanf("%d%d",&k,&cur);
        forest.insert(cur);
        for(int j=1;j<k;++j){
            scanf("%d",&birds);
            forest.insert(birds);
            Union(cur, birds);
        }
    }
    int cnt = 0;
    for(auto &bird:forest){
        int fa = findfather(bird);
        if(isTree[fa] == false){
            isTree[fa] = true;
            cnt++;
        }
    }
    printf("%d %d\n",cnt,(int)forest.size());
    scanf("%d",&k);
    for(int i=0;i<k;++i){
        int a,b;
        scanf("%d%d",&a,&b);
        int fa = findfather(a);
        int fb = findfather(b);
        if(fa == fb) printf("Yes\n");
        else printf("No\n");
    }
}
